vector/src/main/java/im/vector/app/features/location/LocationTracker.kt

```diff
@@ -117,34 +151,84 @@ class LocationTracker @Inject constructor(
-            Timber.d("## LocationTracker. onLocationChanged: $location")
+            Timber.d("onLocationChanged: $location")
         } else {
-            Timber.d("## LocationTracker. onLocationChanged: ${location.provider}")
+            Timber.d("onLocationChanged: ${location.provider}")
         }
-        notifyLocation(location, isLive = true)
-    }
 
-    private fun notifyLocation(location: Location, isLive: Boolean) {
         when (location.provider) {
+            LocationManager.FUSED_PROVIDER -> {
+                hasLocationFromFusedProvider = true
+            }
             LocationManager.GPS_PROVIDER -> {
-                hasGpsProviderLiveLocation = isLive
+                if (hasLocationFromFusedProvider) {
+                    hasLocationFromGPSProvider = false
+                    // Ignore this update
+                    Timber.d("ignoring location from ${location.provider}, we have location from fused provider")
+                    return
+                } else {
+                    hasLocationFromGPSProvider = true
+                }
             }
             else -> {
-                if (hasGpsProviderLiveLocation) {
+                if (hasLocationFromFusedProvider || hasLocationFromGPSProvider) {
                     // Ignore this update
-                    Timber.d("## LocationTracker. ignoring location from ${location.provider}, we have gps live location")
+                    Timber.d("ignoring location from ${location.provider}, we have location from GPS provider")
                     return
                 }
             }
         }
+
+        debouncer.debounce(LOCATION_DEBOUNCE_ID, MIN_TIME_TO_UPDATE_LOCATION_MILLIS) {
+            notifyLocation(location)
+        }
+    }
+
+    private fun notifyLocation(location: Location) {
+        if (BuildConfig.LOW_PRIVACY_LOG_ENABLE) {
+            Timber.d("notify location: $location")
+        } else {
+            Timber.d("notify location: ${location.provider}")
+        }
+
         val locationData = location.toLocationData()
         lastLocation = locationData
-        callbacks.forEach { it.onLocationUpdate(locationData) }
+        onLocationUpdate(locationData)
     }
 
     override fun onProviderDisabled(provider: String) {
-        Timber.d("## LocationTracker. onProviderDisabled: $provider")
-        callbacks.forEach { it.onLocationProviderIsNotAvailable() }
+        Timber.d("onProviderDisabled: $provider")
+        when (provider) {
+            LocationManager.FUSED_PROVIDER -> hasLocationFromFusedProvider = false
+            LocationManager.GPS_PROVIDER -> hasLocationFromGPSProvider = false
+        }
+
+        locationManager?.allProviders
+                ?.takeIf { it.isEmpty() }
+                ?.let {
+                    Timber.e("all providers have been disabled")
+                    onNoLocationProviderAvailable()
+                }
+    }
+
+    private fun onNoLocationProviderAvailable() {
+        callbacks.forEach {
```

Comment:
Is there a risk of `ConcurrentModificationException` here?

Reply:
Hum I guess there is... I will try to add a test to see if it passes.

Reply:
Updated here: [1ee852d](https://github.com/vector-im/element-android/pull/5923/commits/1ee852dbf4e223ebab5a7de775ba4e2c14de3b6c)